junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TypeBasedParameterResolver.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2015-2019 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.jupiter.engine.extension;
+
+import java.lang.reflect.Parameter;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+
+/**
+ * {@link ParameterResolver} adapter which resolve a parameter based on its exact type.
+ * @param <T> the type of the parameter to resolve
+ * @since 5.6
+ */
+public abstract class TypeBasedParameterResolver<T> implements ParameterResolver {
+
+	/**
+	 * Resolve an argument for the {@link Parameter} in the supplied {@link ParameterContext}
+	 * for the supplied {@link ExtensionContext}.
+	 *
+	 * <p>This method is only called by the framework if the type argument {@code T}
+	 * matches the type of the {@link Parameter} in the {@link ParameterContext}.
+	 *
+	 * <p>The {@link java.lang.reflect.Method} or {@link java.lang.reflect.Constructor}
+	 * in which the parameter is declared can be retrieved via
+	 * {@link ParameterContext#getDeclaringExecutable()}.
+	 *
+	 * @param parameterContext the context for the parameter for which an argument should
+	 * be resolved; never {@code null}
+	 * @param extensionContext the extension context for the {@code Executable}
+	 * about to be invoked; never {@code null}
+	 * @return the resolved argument for the parameter of type {@code T}
+	 * @see #supportsParameter
+	 * @see ParameterResolver#resolveParameter(ParameterContext, ExtensionContext)
+	 * @see ParameterContext
+	 */
+	abstract protected T resolve(ParameterContext parameterContext, ExtensionContext extensionContext)
+			throws ParameterResolutionException;
+
+	@Override
+	public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+			throws ParameterResolutionException {
+		return getParameterType(parameterContext).equals(enclosedTypeOfParameterResolver());
+	}
+
+	@Override
+	public final Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+			throws ParameterResolutionException {
+		return resolve(parameterContext, extensionContext);
+	}
+
+	private Type getParameterType(ParameterContext parameterContext) {
+		return parameterContext.getParameter().getParameterizedType();
+	}
+
+	private Type enclosedTypeOfParameterResolver() {
+		return ((ParameterizedType) this.getClass().getGenericSuperclass()).getActualTypeArguments()[0];
```

Comment:
:x: This won't work when there's an intermediate class between the concrete implementation and `TypeBasedParameterResolver`.

Reply:
I changed the code to look into the hierarchy of the class and also implemented a test that illustrates the limitation you found.

Let me know if this is better.

Reply:
Yep, that looks better. Thanks!